Add annotation to show logs of prerequisite pods

rata opened the pull request:
* Adds annotation to show logs of executed prerequisite pods

Fixes: #3003 

@grosser does this make sense to you?

The only disadvantage I see is that the annotation is on the pod, not the job, so you need one annotation for prerequisite at the job kind and another at the pod spec inside the job.

I mean, something like:

```
apiVersion: batch/v1
kind: Job
metadata:
  annotations:
    samson/prerequisite: "true"
  name: data-migration
  labels:
    project: project
    role: data-migration
spec:
  template:
    metadata:
      labels:
        project: project
        role: data-migration
      annotations:
        samson/show_logs_on_success: "true"
    spec:
      restartPolicy: Never
      ...
```

I tried first using an env var, when I had that working I tried the annotation on the pod. But then I realized it's on the pod, not the job or something else, so tried to do it on the job. But not sure that part of the code can check easily for that. I think it's kind of tricky, as it needs to check for prerequisites (not sure that part of the code knows about them), but then check for annotations there and print for all pods that belong to that resource (following the owner reference, hopefully, but not sure that works fine in Kubernetes. On not sure on which versions)

Using just an annotation for the pod seems simpler, now that I think about it. But will try to look at adding the annotation at the job level.  

What do you think? Let me know if at the pod level it's fine (seems simpler, but not sure :))

/cc @zendesk/samson

### Tasks
 - [ ] :+1: from team

### References
 - Jira link:

### Risks
- Level: Low

rata committed:
Add annotation to show logs during deploy

grosser commented on plugins/kubernetes/app/models/kubernetes/deploy_executor.rb:
```diff
@@ -392,13 +406,16 @@ def deploy(release_docs)
     end
 
-    def deploy_and_watch(release_docs, timeout)
+    def deploy_and_watch(release_docs, timeout, show_logs_on_success)
```
show_logs_on_success is not a logical 3rd argument, so should use keyword args

`, show_logs_on_success:` and then call with `show_logs_on_success: true/false`

grosser replied:
also the flag does not make it show logs, just 'maybe' show logs, so maybe better to use `prerequisite: true`

grosser replied:
... we can also show logs for pods that are not prerequisites if the user asked for them ... so don't need this flag at all

rata replied:
Sure! I'll change it to keyword args, thanks for the tips (really valuable for non rubyst :))

Do you think we don't need a third flag? Or a flag to only show prerequisites and show them all (i.e. without the pod.completed? filter)?

I changed it to the latter. The reason is that if we print the same after the deploy (not prerequisites, of deploys deploys) ends, it will still find the logs of the prerequisites pods and it will be printed again. I'm not sure how to filter those, nor if it's worth it.

But if we do the print after the deploys ends, without any more filters, then if something was print after the prerequisites run, then it will be printed again. And that seems confusing, it's just a duplicated print (most of the times).

Maybe after the deploys deploys, we can print but excluding pods that were completed when we printed the first time. But it seems a little messier than I'd like, and not sure the functionality it's worth the complexity.

Also, after the deploys ends, most of the time you deploy some always living objects (like deployments). The ones that are short-lived are probably printed in the first interruption, after prerequisites deploys, so there is no much gain, I think, printing after the deploy ends again.

What do you think?

grosser commented on plugins/kubernetes/app/models/kubernetes/api/pod.rb:
```diff
@@ -18,6 +18,10 @@ def namespace
       end
 
+      def annotations
+        @pod.dig(:metadata, :annotations)
```
needs to be `||= {}` or it will blow up when there are no annotations
```suggestion
        @pod[:metadata][:annotations] ||= {}
```

rata replied:
Good catch, thanks! Applied and squashed now :)

grosser commented on plugins/kubernetes/app/models/kubernetes/deploy_executor.rb:
```diff
@@ -392,13 +406,16 @@ def deploy(release_docs)
           finish_blue_green_deployment(blue_green)
         end
+        if show_logs_on_success
+          show_pods_logs
```
maybe even `show_pods_logs_if_requested`

rata replied:
Yeah, makes sense! I was in doubt, just because of this (it maybe shows logs), but didn't come up with a better name. Thanks!

I changed the func to that name, and changed the variable to `show_logs_if_requested`.

Now the conditional on the var name and on the function name make me doubt, but well... naming is hard. Let me know what you think :)

grosser commented on plugins/kubernetes/app/models/kubernetes/deploy_executor.rb:
```diff
@@ -142,6 +142,20 @@ def fetch_pods
+      log_end_time = 0.seconds.from_now
+      pods = fetch_pods
+      pods.each do |pod|
+        if pod.completed? && pod.annotations[:'samson/show_logs_on_success']
```
better check for
`pod.annotations[:'samson/show_logs_on_success'] == 'true'`

grosser replied:
... since showing logs has a timeout, there is no real reason to not allow it for running pods if the user requested it

rata replied:
Ohh, it makes sense! I removed that check! :)

grosser commented on plugins/kubernetes/app/models/kubernetes/deploy_executor.rb:
```diff
@@ -142,6 +142,20 @@ def fetch_pods
+          @output.puts "\n------------------------------------------\n"
+        end
+      end
+    rescue
```
```suggestion
    rescue StandardError
```

grosser commented:
needs an actual test of `it "shows pod logs when requested"`

grosser commented:
let me know if you get stuck, I can help out or take it over if it gets to funky :)

rata committed:
Add tests

rata commented:
@grosser thanks a lot for the review! And the help. I think I managed to address almost all of the comments. Added some tests too, but please check :)

I'm waiting to see if tests pass correctly now :-D

rata committed:
WIP

rata commented:
@grosser so, tests are not working. If you can help me, it would be awesome.

I modified the `batch/v1` job, adding the annotation (IIUC, without typos), and the output from pods is not there. I added some prints to debug (now on this PR, but just to show) and I really don't see why. For some reason, the pod is not having the pod annotation. But not sure why.

Running samson on my own and manually doing a deploy of a job with the annotation works. So I'm not sure I'm missing something really silly (like some typo or something obvious) or just something that changes in the tests.

Any help is appreciated :-)

rata commented:
My guess is that in the simulated environment for the test, something is not working. Maybe fetch_pods nor returning correctly or for some reason the pod not having the annotation.

I think is probably I don't know how to use the test infrastructure :-/

grosser commented:
I think you found a fundamental bug, fetch_pods should only fetch pods from given release-docs
it never caused a bug because we don't fail if more than requested pods are created, which is also a bug :D 
... I'll fix these and than this PR should be simpler :)

grosser commented:
ah that never was a problem since release_statuses selects filters by role_id

grosser commented:
taking this over now, stay tuned
